Tutorial (#9)
reactjs/zh-hant.reactjs.org: content/tutorial/tutorial.md

  - "docs/tutorial-zh-CN.html"
---

你不需要有任何 React 的基礎知識就能使用這份學習指南。

## 在我們開始這份教學課程之前 {#before-we-start-the-tutorial}

在這份學習指南中，我們會練習做一個小遊戲。**也許你會很想跳過這份指南，因為你不是遊戲開發者 -- 但請試著跟著做做看。** 在這份學習指南中，你所學到的技術是你做任何 React 應用程式的基礎，掌握基礎後會讓你對 React 有更深入的了解。

>Tip
>
>這份學習指南是設計給那些喜歡**從做中學**的人們。如果你比較喜歡從零開始學習概念的話，請參考我們的[逐步教學](/docs/hello-world.html)。你可能會發現本篇指南和教學課程其實是相輔相成的。

這份指南分成以下幾個部分：

* [教學設定](#setup-for-the-tutorial)會給你一個開始這份指南的**起始點**。
* [概論](#overview)會教你 React 中的**重要基礎**：component、prop 和 state。
* [完成遊戲](#completing-the-game)會教你在 React 開發中**最常見的技術**。
* [加入時間旅行](#adding-time-travel)會讓你對 React 獨特的優點有**更深的見解**。

你不需要一次就完成所有的部分才能從這份學習指南中獲益。試著練習越多越好 -- 即使只有一兩個部分。

當你在跟著指南練習時，複製貼上程式碼是沒關係的，但我們建議你把程式自己寫過一遍。這會幫助你訓練手感並加強理解。

### 我們要做什麼？ {#what-are-we-building}

在這份學習指南中，我們會教你如何用 React 做一個互動式的圈圈叉叉小遊戲。

在這裡你可以看到我們將會做什麼：**[完成結果](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**。如果你看不懂其中的程式碼，或是你對其中的語法不熟，請不要擔心！這份指南的目的就是要幫助你了解 React 及其語法。

我們建議你在往下看這份指南之前，先看看上面這個圈圈叉叉小遊戲。你會注意到遊戲格子的右邊有一個數字列表。這個列表將會列出遊戲中的所有動作的歷史，並在遊戲進行的同時更新。

當你稍微了解這個圈圈叉叉小遊戲是怎麼玩的之後，你就可以把它關掉了。 在這份指南中，我們會從更簡單的模板開始。我們的下一步是進行設定以幫助你開始開發這個遊戲。

### 先決條件 {#prerequisites}

> 我們假設你對 HTML 和 JavaScript 有一定的熟悉度，但即使你的背景是另一種程式語言，你應該也能游刃有餘地理解這份指南。我們也假設你對程式語言的中的某些概念，如 function、 object、array 以及（某種程度上）class，有一定的涉獵。

Reviewer: ```suggestion
我們假設你對 HTML 和 JavaScript 有一定的熟悉度，但即使你的背景是另一種程式語言，你應該也能游刃有餘地理解這份指南。我們也假設你對程式語言的中的某些概念，如 function、object、array 以及（某種程度上）class，有一定的涉獵。
```

如果你需要複習 JavaScript，我們建議你閱讀這份[教學指南](https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript)。請注意我們也會用到一些 ES6，也就是 JavaScript 最新的版本之一。在這份指南中，我們將會使用 [arrow function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)、[class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)、[`let`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let)，和 [`const`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) 等表達式。你也可以用 [Babel REPL](babel://es5-syntax-example) 看看 ES6 的程式碼是如何被編譯的。

## 教學設定 {#setup-for-the-tutorial}

完成這份指南有兩種方法：你可以在瀏覽器中寫程式碼，或在你的電腦裡建立一個本地的開發環境。

### 設定選項 1：在瀏覽器中寫程式碼 {#setup-option-1-write-code-in-the-browser}

> 這是開始最快的方法！

Reviewer 2: i suggest change into
> 這是開始的最快方式

or
> 這是最快的開始方式

  - "docs/tutorial-zh-CN.html"
---

你不需要有任何 React 的基礎知識就能使用這份學習指南。

> ## 在我們開始之前 {#在我們開始之前}

Reviewer 3: @chloewlin you don't need to translate anchor link.

Reference: [Universal Style Guide - Heading IDs](https://github.com/reactjs/reactjs.org-translation/blob/master/style-guide.md#heading-ids)

Author: @neighborhood999 Oops! Ok I will fix all anchor links now and push it soon.

在這份學習指南中，我們會練習做一個小遊戲。 **也許你會很想跳過這份指南，因為你不是遊戲開發者 -- 但請試著跟著做做看。** 這份學習指南中你所會學到的技術是你做任何 React App 的基礎， 掌握好基礎會讓你對 React 有更深的了解。

> >溫馨提示

Author: OK. Let's use tip!

>
>這份學習指南是設計給那些喜歡**從做中學**的人們。如果你比較喜歡從零開始學習概念的話，請參考我們的[逐步教學](/docs/hello-world.html)。這份指南跟逐步教學兩者是相輔相成的。

> 這份指南分成以下幾個部分：

Author: I prefer `教學指南` and `學習指南`. I try to use them alternatively so that our doc doesn't sound boring. :D

* [教學設定](#setup-for-the-tutorial) 會給你一個開始這份指南的**起始點**。
* [概論](#概論) 會教你 React 中的**重要基礎**： components，props，和 state。
* [完成遊戲](#完成遊戲) 會教你在 React 開發中**最常見的技術**。
* [加入 Time Travel](#加入Time-Travel) 會讓你對 React 獨特的優點有**更深的見解**。

你不需要一次就完成所有的部分才能從這份學習指南中獲益。試著練習越多越好 -- 即使只有一兩個部分。

當你在跟著指南練習時，複製貼上程式碼是沒關係的，但我們建議你把程式自己寫過一遍。 這會幫助你訓練手感並加強理解。

### 我們要做什麼？ {#what-are-we-building?}

在這份指南中，我們會教你如何用 React 做一個互動式的圈圈叉叉小遊戲。

> 在這裡你可以看到我們將會做什麼： **[最終成果](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**。 如果你看不懂其中的程式碼，或是你對其中的語法不熟，請不要擔心。

Reviewer 3: ```suggestion
在這裡你可以看到我們將會做什麼：**[完成結果](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**。如果你看不懂其中的程式碼，或是你對其中的語法不熟，請不要擔心！這份指南的目的就是要幫助你了解 React 及其語法。
```

> 這份指南的目的就是要幫助你了解 React 及其語法。

Reviewer 3: This line can delete.

> 我們建議你在往下看這份指南前先看看上面這個圈圈叉叉小遊戲。你會注意到遊戲格子的右邊有一個數字列表。這個列表將會列出遊戲中的所有動作的歷史，並在遊戲進行的同時更新。

Reviewer 3: ```suggestion
我們建議你在往下看這份指南之前，先看看上面這個圈圈叉叉小遊戲。你會注意到遊戲格子的右邊有一個數字列表。這個列表將會列出遊戲中的所有動作的歷史，並在遊戲進行的同時更新。
```

當你稍微了解這個圈圈叉叉小遊戲是怎麼玩的之後，你就可以把它關掉了。 在這份指南中，我們會從更簡單的模板開始。我們的下一步是進行設定以幫助你開始開發這個遊戲。

### 先決條件 {#prerequisites}

> 我們假設你對 HTML 和 JavaScript 有一定的熟悉度，但即使你的背景是另一種程式語言，你應該也能游刃有餘地理解這份指南。我們也假設你對程式語言的中的某些概念，如 functions 、 objects 、 arrays 以及（某種程度上）classes ，有一定的涉獵。

Reviewer 3: ```suggestion
我們假設你對 HTML 和 JavaScript 有一定的熟悉度，但即使你的背景是另一種程式語言，你應該也能游刃有餘地理解這份指南。我們也假設你對程式語言的中的某些概念，如 function、 object、array 以及（某種程度上）class，有一定的涉獵。
```

> 如果你需要複習 JavaScript ，我們建議你閱讀這份[教學指南](https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript)。請注意我們也會用到一些 ES6，也就是JavaScript 最新的版本之一。在這份指南中，我們將會使用 [arrow functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)，[classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)，[`let`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let)， 和 [`const`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) 等表達式。你也可以用 [Babel REPL](babel://es5-syntax-example) 看看 ES6 的程式碼是如何被編譯的。

Reviewer 3: ```suggestion
如果你需要複習 JavaScript，我們建議你閱讀這份[教學指南](https://developer.mozilla.org/en-US/docs/Web/JavaScript/A_re-introduction_to_JavaScript)。請注意我們也會用到一些 ES6，也就是 JavaScript 最新的版本之一。在這份指南中，我們將會使用 [arrow function](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions)、[class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes)、[`let`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/let)，和 [`const`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const) 等表達式。你也可以用 [Babel REPL](babel://es5-syntax-example) 看看 ES6 的程式碼是如何被編譯的。
```

## 教學設定 {#setup-for-the-tutorial}

完成這份指南有兩種方法：你可以在瀏覽器中寫程式碼，或在你的電腦裡建立一個本地的開發環境。

### 設定選項 1：在瀏覽器中寫程式碼 {#setup-option-1-write-code-in-the-browser}

這是開始最快的方法！

> 首先，在瀏覽器中的分頁中打開這份**[Starter Code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)**。 你應該會看到一個空白的圈圈叉叉遊戲格和一些 React 的程式碼。在接下來的指南中，我們會修改中的 React 程式。

Reviewer 3: ```suggestion
首先，在瀏覽器中的分頁中打開這份 **[Starter Code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)**。你應該會看到一個空白的圈圈叉叉遊戲格和一些 React 的程式碼。我們會在接下來的指南中，修改 React 的程式碼。
```

> 你現在可以跳過第二個選項，然後直接看[概論](#概論)的部分並瞭解 React 如何運作。

Reviewer 3: ```suggestion
你現在可以跳過第二個選項，並前往[概論](#overview)的章節來了解 React 如何運作。
```

### 設定選項 2：建立本地開發環境 {#setup-option-2-local-development-environment}

對完成這份指南來說，這完全是看你是否需要，而非必要的設定！

<br>

<details>

> <summary><b>非必要: 以下為在本地環境中使用你想要的編輯器所需要的步驟</b></summary>

Reviewer 3: ```suggestion
<summary><b>可選則：以下為在本地環境中使用你想要的編輯器所需要的步驟</b></summary>
```

這個設定需要花些時間，但能讓你在你想要的編輯器上完成這份指南。以下是設定步驟：

1. 確認你有安裝一個較新的版本的 [Node.js](https://nodejs.org/en/)。

> 2. 按照[Create React App 安裝步驟](/docs/create-a-new-react-app.html#create-react-app)的指示設立一個新專案。

Reviewer 3: ```suggestion
2. 按照 [Create React App 安裝步驟](/docs/create-a-new-react-app.html#create-react-app)的指示建立一個新專案。
```

```bash
npx create-react-app my-app
```

3. 刪除新專案中 `src/` 資料夾裡的所有檔案。

> 注意：
>

> >**請不要刪除整個 `src` 資料集，只要刪除裡面的原始檔就好。** 下一步，我們將會把內建的原始檔換成這個遊戲所需的範例檔。

Reviewer 3: ```suggestion
>**請不要刪除整個 `src` 資料夾，只要刪除裡面的原始檔就好。**下一步，我們將會把預設的原始檔換成這個遊戲所需的範例檔。
```

```bash
cd my-app
cd src

# 如果你用的是 Mac 或 Linux：
rm -f *

# 如果你用 Windows：
del *

# 然後，回到原本的專案資料夾
cd ..
```

> 4. 在 `src/` 資料夾中加入命名為 `index.css` 的檔案，用[這個CSS](https://codepen.io/gaearon/pen/oWWQNa?editors=0100)。

Reviewer 3: ```suggestion
4. 在 `src/` 資料夾中加入[這個 CSS 程式碼](https://codepen.io/gaearon/pen/oWWQNa?editors=0100)並命名為 `index.css`。
```

> 5. 在 `src/` 資料夾中加入命名為 `index.js` 的檔案，用[這個JS](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)。

Reviewer 3: ```suggestion
5. 在 `src/` 資料夾中加入[這個 JS 程式碼](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)並命名為 `index.js` 的檔案。
```

6. 在 `src/` 資料夾中加入這三行程式碼在 `index.js` 檔案的最上方：

```js
import React from 'react';
import ReactDOM from 'react-dom';
import './index.css';
```

> 現在如果你在專案資料夾跑 `npm start` 並在瀏覽器中打開 `http://localhost:3000` ， 你會看到一個空白的圈圈叉叉遊戲格。

Reviewer 3: ```suggestion
現在如果你在專案資料夾執行 `npm start` 並在瀏覽器中打開 `http://localhost:3000`，你會看到一個空白的圈圈叉叉遊戲格。
```

> 我們推薦你跟著 [這份指南](https://babeljs.io/docs/editors/) 去設定你的編輯器中的 syntax highlighting。

Reviewer 3: ```suggestion
我們推薦你跟著[這份指南](https://babeljs.io/docs/editors/)來設定你編輯器中的 syntax highlighting。
```

`syntax highlighting` -> `語法高亮`?

</details>

### 救命呀，我卡住了！ {#help-im-stuck}

> 如果你卡住了，請看看 [社群資源](/community/support.html)。 其中 [Reactiflux Chat](https://discord.gg/0ZcbPKXt5bZjGY5n) 可以讓你很快得到幫助。如果沒人回你，或你還是卡住了，請填寫一個 issue，我們會提供協助。

Reviewer 3: ```suggestion
如果你卡住了，請參考[社群資源](/community/support.html)。其中 [Reactiflux Chat](https://discord.gg/0ZcbPKXt5bZjGY5n) 可以讓你很快得到幫助。如果沒人回覆，或者你依然無法解決，請填寫一個 issue，我們會提供協助。
```

## 概論 {#overview}

設定完成後，讓我們來討論 React 的概論吧！

### React 是什麼？ {#what-is-react}

> React 是一個宣告式、高效且很有彈性的 JavaScript 函式庫，用以建立使用者介面。它讓你用小巧而獨立，名為「components」的程式碼，建立複雜的 UI。

Reviewer 3: ```suggestion
React 是一個宣告式、高效且具有彈性的 JavaScript 函式庫，用以建立使用者介面。它讓你使用小巧而獨立的「component」，來建立複雜的 UI。
```

`宣告式` -> `陳述式`?

> React 有數種不同的 components，但我們先從 `React.Component` 這個 subclasses 開始：

Reviewer 3: ```suggestion
React 有數種不同的 component，但我們先從 `React.Component` 這個 subclass 開始：
```

// Example usage: <ShoppingList name="Mark" />
```

> 我們稍後會解釋那些看起來很有趣像是 XML 的 tags。 我們使用 components 告訴 React 我們想要在螢幕上看到什麼。當我們的資料改變時，React 將會很有效率地更新並 re-render 我們的 components。

Reviewer 3: ```suggestion
我們稍後會解釋那些看起來很有趣像是 XML 的 tag。我們使用 component 告訴 React 我們想要在螢幕上看到什麼。當我們的資料改變時，React 將會很有效率地更新並 re-render 我們的 component。
```

> 在這裡，ShoppingList 是一個 **React 的 component class**，或 **React component type**。 Component 會接受名為 `props` 的參數 ( props 為 "properties" 的簡稱)，並透過 `render` 這個方法回傳一個有階層架構的 views 到螢幕上。

Reviewer 3: ```suggestion
在這裡，ShoppingList 是一個 **React 的 component class**，或 **React component type**。Component 會接受名為 `props` 的參數（「properties」的簡稱），並透過 `render` 這個方法回傳一個有階層架構的 view 到螢幕上。
```

> 名為 `render` 的方法回傳你想在螢幕上看到的*描述（ description ）*。 React 接收這個描述並展示其結果。其中，`render` 回傳的是一個 **React 元素**，也就是一個輕量的、包含該 render 什麼的描述。大部分 React 的開發者會使用一種特殊的、名為「 JSX 」的語法，因為它讓這些架構寫起來更容易。 `<div />` 語法在 build time 時被建立為 `React.createElement('div')`。上述的例子其實也等同於以下這些程式碼：

Reviewer 3: ```suggestion
`render` 方法回傳你想在螢幕上看到的*描述*。React 接收這個描述並展示其結果。其中，`render` 回傳的是一個 **React element**，也就是一個 render 內容的輕量描述。大部分 React 的開發者會使用一種特殊的語法，被稱為「JSX」，因為它讓這些結構寫起來更容易。`<div />` 語法在構建時被建立為 `React.createElement('div')`。上述的例子其實也等同於以下這些程式碼：
```

);
```

[你可以按這裡看到完整版](babel://tutorial-expanded-version)

> 如果你想了解更多 `createElement()`，在 [API 參考](/docs/react-api.html#createelement)中有更詳細的解釋，但我們不會在這份指南中用到它。我們會繼續使用 JSX。

Reviewer: ```suggestion
如果你想了解更多 `createElement()`，在 [API 參考指南](/docs/react-api.html#createelement)中有更詳細的解釋，但我們不會在這份指南中用到它。我們會繼續使用 JSX。
```

Same as last line in this page.

);
```

[你可以按這裡看到完整版](babel://tutorial-expanded-version)。

> 如果你想了解更多， `createElement()` 在 [API 參考](/docs/react-api.html#createelement)中有更詳盡的解釋，但我們不會在這份指南中用到它。我們會繼續使用 JSX。

Reviewer 3: ```suggestion
如果你想了解更多 `createElement()`，在 [API 參考](/docs/react-api.html#createelement)中有更詳細的解釋，但我們不會在這份指南中用到它。我們會繼續使用 JSX。
```

> JSX 就跟 JavaScript ㄧ樣強大。 你可以在 JSX 中的括號中放入 *任何* JavaScript 的表達式。 每個 React 元素都是一個 JavaScript 的物件，你可以把它存在一個變數中或在你的程式中傳來傳去。

Reviewer 3: ```suggestion
JSX 就跟 JavaScript ㄧ樣強大。你可以在 JSX 中的括號中放入*任何* JavaScript 的表達式。每個 React element 都是一個 JavaScript 的 object，你可以把它存在一個變數中或者在程式中互相傳遞。
```

> 上面的 `ShoppingList` component 只會 renders 內建如 `<div />` 和 `<li />` 的 DOM components。 但你也可以建立並 render 客製的 React components。例如，我們現在可以用 `<ShoppingList />` 來使用 shopping list。每個 React component 都是封裝（encapsulated）好並能獨立運作的。 React 讓你能用簡單的 components 建立複雜的 UI。

Reviewer 3: ```suggestion
上述的 `ShoppingList` component 只會 render 內建如 `<div />` 和 `<li />` 的 DOM component。此外，你也可以組合 render 客製的 React component。例如，我們現在可以用 `<ShoppingList />` 來使用 shopping list。每個 React component 都是封裝好並能獨立運作的。React 讓你能用簡單的 component 建立複雜的 UI。
```

## 檢查你的 Starter Code {#inspecting-the-starter-code}

> 如果你打算在 **你的瀏覽器** 中練習這份指南的話，請在分頁中開啟這份程式：**[Starter Code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)**。如果你打算在**本地開發環境** 中練習，在你的專案資料夾中打開 `src/index.js` （你在[setup](#setup-option-2-local-development-environment)時已接觸過這個檔案）。

Reviewer 3: ```suggestion
如果你打算在**你的瀏覽器** 中練習這份指南的話，請在分頁中開啟這份程式：**[Starter Code](https://codepen.io/gaearon/pen/oWWQNa?editors=0010)**。如果你打算在**本地開發環境** 中練習，在你的專案資料夾中打開 `src/index.js`（你在 [setup](#setup-option-2-local-development-environment) 時已接觸過這個檔案）。
```

> 這個 Starter Code 是我們接下來練習的基礎。我們已幫你準備好了CSS，如此一來，你只需要把注意力集中在學習 React 和完成圈圈叉叉小遊戲的程式上。

Reviewer 3: ```suggestion
這個 Starter Code 是我們接下來練習的基礎。我們已幫你準備好了 CSS，如此一來，你只需要把注意力集中在學習 React 和完成圈圈叉叉小遊戲的程式上。
```

> 請看看其中的程式。你會注意到我們有三個 React components：

Reviewer 3: ```suggestion
請看看其中的程式。你會注意到我們有三個 React component：
```

* Square
* Board
* Game

> Square component 會 renders 一個按鈕 `<button>`，而 Board 則 render 九個方格。Game component 則是 render 一個完整的遊戲格盤與 placeholder ，這些值我們稍候會修改。目前為止，沒有一個 component 是互動式的。

Reviewer 3: ```suggestion
Square component 會 render 一個按鈕 `<button>`，而 Board 則 render 九個方格。Game component 則是 render 一個完整的遊戲格盤與 placeholder，這些值我們稍候會修改。目前為止，沒有一個 component 是互動式的。
```

}
```

修改前：

![React Devtools](../images/tutorial/tictac-empty.png)

修改後：你應該會看到每個方格中都有一個數字。

![React Devtools](../images/tutorial/tictac-numbers.png)

**[按這裡看目前的程式碼](https://codepen.io/gaearon/pen/aWWQOG?editors=0010)**

> 恭喜！你剛剛已順利將把 prop 從 parent Board component 傳給 child Square component。傳遞 props (passing props）是 React 的應用程式中資訊從 parent 傳給 children 的方式。

Reviewer 3: ```suggestion
恭喜！你剛剛已順利將把 prop 從 parent Board component 傳給 child Square component。傳遞 props 是 React 的應用程式中資訊從 parent 傳給 children 的方式。
```

}
```

現在，當我們點擊任何一個 Square 時，我們應該能在瀏覽器中收到一個 alert。

>注意：
>

> >為了節省空間和避免 [`this` 令人困惑的行為](https://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/)，我們將會在這裡以及以下的指南使用[arrow function 語法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) 來寫 event handlers：

Reviewer 3: ```suggestion
>為了節省空間和避免 [`this` 令人困惑的行為](https://yehudakatz.com/2011/08/11/understanding-javascript-function-invocation-and-this/)，我們將會在這裡以及以下的指南使用 [arrow function 語法](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions) 來寫 event handler：
```

>}
>```
>

> >注意在 how with `onClick={() => alert('click')}`中，我們會把一個 *function* 做為 `onClick` 的 prop往下傳。這個 function 只會在被點擊後觸發。 把 `() =>` 寫成 `onClick={alert('click')}` 是一個常見的錯誤，這會造成 component 在每次 re-render 時都觸發 alert。

Reviewer 3: ```suggestion
>注意在 how with `onClick={() => alert('click')}`中，我們會把一個 *function* 做為 `onClick` 的 prop 往下傳。這個 function 只會在被點擊後觸發。把 `() =>` 寫成 `onClick={alert('click')}` 是一個常見的錯誤，這會造成 component 在每次 re-render 時都觸發 alert。
```

> 下一步，我們要讓 Square component 記得它被點擊了，並在方格中填入 X 這個記號。Components 使用 **state** 來保持記憶。

Reviewer 3: ```suggestion
下一步，我們要讓 Square component 記得它被點擊了，並在方格中填入 X 這個記號。Component 使用 **state** 來保持狀態。
```

}
```

>注意：
>

> >在 [JavaScript classes](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) 中， 當你定義一個 subclass 的 constructor 時，你總是會需要呼叫 `super`。所有的 React component class ，凡是有 `constructor` 的，都應該要從呼叫 `super(props)` 開始。

Reviewer 3: ```suggestion
>在 [JavaScript class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Classes) 中，當你定義一個 subclass 的 constructor 時，你總是會需要呼叫 `super`。所有的 React component class，凡是有 `constructor` 的，都應該要從呼叫 `super(props)` 開始。
```

}
```

藉由從 Square 的 `render` 方法中的 `onClick` handler 呼叫 `this.setState`，我們告訴 React：當該 Square 的 `<button>` 被點擊時，要 re-render。在這個修改後， Square 的 `this.state.value` 值將會變成 `'X'` ，所以我們將會在遊戲格盤中看到 `'X'`。當你點擊任何一個方格，應該能看到 `'X'`。

> 當你在一個 component 中呼叫 `setState` 時， React 也會自動更新其中的 child component。

Reviewer 3: ```suggestion
當你在一個 component 中呼叫 `setState` 時，React 也會自動更新其中的 child component。
```

**[按這裡看目前的程式碼](https://codepen.io/gaearon/pen/VbbVLg?editors=0010)**

### 開發者工具 {#developer-tools}

> React 在 [Chrome](https://chrome.google.com/webstore/detail/react-developer-tools/fmkadmapgofadopljbjfkapdkoienihi?hl=en) 和 [Firefox](https://addons.mozilla.org/en-US/firefox/addon/react-devtools/) 中的 Devtools extension 讓你用你的瀏覽器的開發者工具檢查 React component 的樹。

Reviewer 3: Should we keep `樹` to `tree`?

Author: Sure!

<img src="../images/tutorial/devtools.png" alt="React Devtools" style="max-width: 100%">

React DevTools 讓你檢查你的 React component 中的 props 和 state。

在安裝 React DevTools 後，你在網頁上對任何元素按右鍵，之後按 Inspect 開啟開發者工具，React 的 tab 會是右邊數來第一個。

**然而，請注意在 CodePen 中使用開發者工具需要一些額外的步驟：**

1. 登入網站，或註冊並確認你的 email (為防止 spam 的必要手續)。
2. 點擊 Fork 按鈕。
3. 點擊 Change View 並選擇 Debug mode。
4. 在新開啟的分頁中，devtools 現在應該有 React 的 tab 了。

## 完成遊戲 {#completing-the-game}

現在我們已經為我們的圈圈叉叉遊戲準備好基本的要件了。為了要做出一個完整的遊戲，我們現在需要在遊戲格盤中交互地放入 X 和 O，我們也需要決定如何分出勝負。

> ### 把 State 向上傳 {#lifting-state-up}

Reviewer 3: ```suggestion
### 將 State 往上傳 {#lifting-state-up}
```

]
```

Board 的 `renderSquare` 方法目前看起來是這樣：

```javascript
  renderSquare(i) {
    return <Square value={i} />;
  }
```

> 一開始，我們從 Board [把 `value` 這個 prop 往下傳](#passing-data-through-props) 並在每一個 Square 中顯示數字 0 到 8。在之前的另一個步驟中，我們 [根據 Square 自身的 state](#making-an-interactive-component)把數字換成 Ｘ。這是為什麼 Square 目前會忽略 Board 傳給它的 `value` prop 的原因。

Reviewer 3: ```suggestion
一開始，我們從 Board [把 `value` 這個 prop 往下傳](#passing-data-through-props) 並在每一個 Square 中顯示數字 0 到 8。在之前的另一個步驟中，我們 [根據 Square 本身的 state](#making-an-interactive-component)把數字換成 Ｘ。這是為什麼 Square 目前會忽略 Board 傳給它的 `value` prop 的原因。
```

> 我們現在又會再使用傳遞 prop 的這個機制。 我們會修改 Board 以告訴每個 Square 它現在的值（`'X'`，`'O'`， 或 `null`）該是什麼。 我們已經在 Board 的 constructor 中定義了 `squares` 這個 array，接下來我們會修改 Board 的 `renderSquare` 方法以讀取這個 array:

Reviewer 3: ```suggestion
我們現在又會再使用傳遞 prop 的這個機制。我們會修改 Board 以告訴每個 Square 它現在的值（`'X'`，`'O'`， 或 `null`）該是什麼。我們已經在 Board 的 constructor 中定義了 `squares` 這個 array，接下來我們會修改 Board 的 `renderSquare` 方法以讀取這個 array：
```

  }
```

最後，我們會修改 Game component 中的 `render` 方法，從總是 render 最後一個動作，改為根據 `stepNumber` 來 render 目前正在被選取的動作：

```javascript{3}
  render() {
    const history = this.state.history;
    const current = history[this.state.stepNumber];
    const winner = calculateWinner(current.squares);

    // 以下不需要改變
```

現在，當我們點擊遊戲歷史中的任何一步，我們應該能夠立刻看到圈圈叉叉遊戲棋盤被更新，並顯示棋盤在那一步發生後的模樣。

**[按這裡看目前的程式碼](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**

### 總結 {#wrapping-up}

恭喜！你現在已經創造了一個圈圈叉叉遊戲，它可以：

* 讓你玩圈圈叉叉
* 顯示哪一個玩家取得勝利
* 在遊戲進行的同時儲存遊戲歷史
* 讓玩家回顧遊戲的歷史，並回到棋盤之前的版本

做得好！希望現在你覺得你對 React 的運作有相當程度的理解。

按這裡看看最終的結果：**[完成結果](https://codepen.io/gaearon/pen/gWWZgR?editors=0010)**。

如果你有額外的時間或想練習你的 React 新技巧，下面是一些你可以改進圈圈叉叉小遊戲的想法，依照程度逐漸增加困難度：

1. 在歷史動作列表中，用（欄，列）的格式來顯示每個動作的位置。
2. 在動作列表中，將目前被選取的項目加粗。

> 3. 改寫 Board，用兩個 loop 創造方格，而非使用 hardcode。

Reviewer: ```suggestion
3. 改寫 Board，使用兩個 loop 創造方格而不是寫死它。
```
